jsm/FlowWebComponent.js

```diff
@@ -0,0 +1,267 @@
+                [ "label", { className: "LabelElement", path: "./elements/LabelElement.js" } ],
+                [ "title", { className: "TitleElement", path: "./elements/TitleElement.js" } ]
+            ] ),
+            input: new Map( [
+                [ "button", { className: "ButtonInput", path: "./inputs/ButtonInput.js" } ],
+                [ "color", { className: "ColorInput", path: "./inputs/ColorInput.js" } ],
+                [ "number", { className: "NumberInput", path: "./inputs/NumberInput.js" } ],
+                [ "select", { className: "SelectInput", path: "./inputs/SelectInput.js" } ],
+                [ "slider", { className: "SliderInput", path: "./inputs/SliderInput.js" } ],
+                [ "string", { className: "StringInput", path: "./inputs/StringInput.js" } ],
+                [ "text", { className: "TextInput", path: "./inputs/TextInput.js" } ],
+                [ "toggle", { className: "ToggleInput", path: "./inputs/ToggleInput.js" } ]
+            ] )
+        };
+
+    }
+
+    add(type, name, desc) {
+
+        this.cache[type].set(name, desc);
+
+    }
+
+    get(type, name) {
+
+        return this.cache[type].has(name) ? this.cache[type].get(name) : this.default[type];
+
+    }
+
+    remove(type, name) {
+
+        this.cache[type].delete(name);
+
+    }
+}
+
+const optionsBuilder = params => {
+
+    let options = { components: new Component() };
+
+    const { width, height, components } = params;
+
+    if ( typeof width === "number" )
+        options.width = toPX( width );
+
+    if ( typeof height === "number" )
+        options.height = toPX( height );
+
+    if ( components ) {
+
+        if ( Array.isArray(components) ) {
+
+            components.forEach( ( { type, name, desc } ) => {
+
+                options.components.add( type, name, desc );
+
+            } );
+
+        } else if ( typeof components === "object") {
+
+            const { type, name, desc } = components;
+            options.components.add( type, name, desc );
+
+        }
+
+    }
+
+    return options;
+
+}
+
+export default class FlowElement extends HTMLElement {
+
+    constructor( options ) {
+
+        super();
+
+        this.options = optionsBuilder( options );
+
+        this.attachShadow({ mode: "open" });
+
+        const holderSheet = new CSSStyleSheet();
+        holderSheet.replace( `
+            :host {
+                position: absolute;
+                width: ${ this.options.width };
+                height: ${ this.options.height };
+            }
+        ` );
+
+        this.shadowRoot.adoptedStyleSheets = [ flowSheet, holderSheet ];
+
+    }
+
+    connectedCallback() {
+
+        this.canvas = new Canvas();
+
+        this.shadowRoot.append( this.canvas.dom );
+
+        FlowElement.onDrop = this.canvas.onDrop.bind( this.ondrop );
+
+        FlowElement.start = this.canvas.start.bind( this.canvas );
```

Comment:
Why the use of static function here? I wonder if this is compatible to multiple canvas?